src/Transfer.js

```diff
@@ -0,0 +1,390 @@
+import './Transfer/types.js'
+import React, { Children, useState } from 'react'
+import propTypes from '@dhis2/prop-types'
+
+import { Actions } from './Transfer/Actions.js'
+import { AddAll } from './Transfer/AddAll.js'
+import { AddIndividual } from './Transfer/AddIndividual.js'
+import { Container } from './Transfer/Container.js'
+import { Filter } from './Transfer/Filter.js'
+import { LeftFooter } from './Transfer/LeftFooter.js'
+import { LeftHeader } from './Transfer/LeftHeader.js'
+import { LeftSide } from './Transfer/LeftSide.js'
+import { PickedOptions } from './Transfer/PickedOptions.js'
+import { RemoveAll } from './Transfer/RemoveAll.js'
+import { RemoveIndividual } from './Transfer/RemoveIndividual.js'
+import { ReorderingActions } from './Transfer/ReorderingActions.js'
+import { RightFooter } from './Transfer/RightFooter.js'
+import { RightSide } from './Transfer/RightSide.js'
+import { SourceOptions } from './Transfer/SourceOptions.js'
+import {
+    addAllSelectableSourceOptions,
+    addIndividualSourceOptions,
+    addKeyToReactOptions,
+    createDoubleClickHandlers,
+    extractPickedReactOptions,
+    filterOptions as defaultFilterOptions,
+    filterOutOptions,
+    getSubsetByFilter,
+    isReorderDownDisabled,
+    isReorderUpDisabled,
+    moveHighlightedPickedOptionDown,
+    moveHighlightedPickedOptionUp,
+    removeAllPickedOptions,
+    removeIndividualPickedOptions,
+    useHighlightedOptions,
+} from './Transfer/helper/index.js'
+import {
+    singleSelectedPropType,
+    multiSelectedPropType,
+} from './common-prop-types.js'
+
+/**
+ * @module
+ * @param {Transfer.PropTypes} props
+ *
+ * @returns {React.Component}
+ *
+ * @example import { Transfer } from @dhis2/ui-core
+ * @see Specification: {@link https://github.com/dhis2/design-system/blob/master/organisms/transfer.md|Design system}
+ * @see Live demo: {@link /demo/?path=/story/transfer--basic|Storybook}
+ */
+export const Transfer = ({
+    onChange,
+
+    children,
+    className,
+    dataTest,
+    disabled,
+    emptyOptionsPlaceholder,
+    emptySelectionPlaceholder,
+    enableOrderChange,
+    filter,
+    filterLabel,
+    filterOptions,
+    height,
+    initialFilter,
+    labelAddAll,
+    labelAddIndividual,
+    labelRemoveAll,
+    labelRemoveIndivitual,
+    leftFooter,
+    leftHeader,
+    maxSelections,
+    optionsWidth,
+    rightFooter,
+    selected,
+    selectedWidth,
+    showFilterInput,
+    onFilterChange,
```

Comment:
Some differences with the Select api here, that I think would be good to align:

* `emptyOptionsPlaceholder` and `emptySelectionPlaceholder`. The select accepts a component under the `empty` prop for this. So it doesn't use the word placeholder.
* `filter`, `filterLabel`, `filterOptions`, `initialFilter` and `showFilterInput`. The Select does not have a modifiable filtering algorithm. It just has a `filterable` prop for toggling the filter and `filterPlaceholder` and `noMatchText` (so different wording as well)
* `labelAddAll`, `labelAddIndividual`, `labelRemoveAll`, `labelRemoveIndivitual` the Select also accepts text for placeholders and buttons, but uses different terminology: `placeholder`, `filterPlaceholder`, `clearText`, `loadingText`, `noMatchText`
* Single component with toggleable single or multiple selections, versus the two exported components for the select.

Now I'm not saying I prefer one over the other, but merely pointing out differences in the API, where the functionality seems similar (though the components are different of course). This could be confusing to devs using both components, so I think we should make sure that we align areas that do similar things, and only diverge where we think it's justified (because for example, the component really needs different functionality). I think it would be good to discuss that.

p.s.: small typo: `labelRemoveIndivitual`

Reply:
> The select accepts a component under the empty prop for this.

`empty` sounds like a modifier, like `dense`. I think we should somehow incorporate the word "default", because it's defaulting to a specified component under a certain condition. The word "placeholder" should probably go as it has a very specific meaning for input elements. But I'm not sure how to name the props then.. `sourceDefaultComponent` & `pickedDefaultComponent`?

> `filter`, `filterLabel`, `filterOptions`, `initialFilter` and `showFilterInput`. The Select does not have a modifiable filtering algorithm. It just has a filterable prop for toggling the filter and filterPlaceholder and noMatchText (so different wording as well)

I first used `enableFilter` which is the equivalent of `filterable`, but I ran into issues once the behavior could be changed.

Let's say you want to add custom components in the header which filter the options (e. g. with a select; by year) but you don't want to show the filter input field. Providing the `filterOptions` callback should suffice to trigger the "filterable" behavior that's inside the component. If it does want to show the filter input, a `filterable` prop wouldn't make much sense, so I chose `showFilterInput`

> `labelAddAll`, `labelAddIndividual`, `labelRemoveAll`, `labelRemoveIndivitual` the Select also accepts text for placeholders and buttons, but uses different terminology: placeholder, filterPlaceholder, clearText, loadingText, noMatchText

I think the word "placeholder" is appropriate in your case. "My" props are "text"s as well. So the question is: `label` vs `text` (tbh I don't care) and prefix vs suffix. I def. prefer the term we choose to prefix the prop as that will group them in the prop types (and ideally in the component's argument list; that's not one of our conventions though).

> Single component with toggleable single or multiple selections, versus the two exported components for the select.

Two separate components is what I tried initially, it increased the complexity so I chose against it. This component does not have a `single` vs `multi` approach, instead the amount of selections can be limited. It currently only allows two values (`1` and `Infinity`), but we can increase the range of allowed selections without introducing a breaking change, we'd just have to extend the UI. It also removed the need to handle two different prop types for `selected`. Bot the range instead of single/multi and the single prop type decrease complexity.

Reply:
`labelRemoveIndivitual` <- fixed

Reply:
> But I'm not sure how to name the props then.. sourceDefaultComponent & pickedDefaultComponent?

Yeah. I think in that case I'd go with `sourceEmptyComponent` and `pickedEmptyComponent`. To me that seems the clearest from a user perspective. Default still seems like it could refer to sth else.

> empty sounds like a modifier, like dense.

Yeah true. Using emptyComponent for the select seems good, if we do the above.

> Let's say you want to add custom components in the header which filter the options (e. g. with a select; by year) but you don't want to show the filter input field. Providing the filterOptions callback should suffice to trigger the "filterable" behavior that's inside the component. If it does want to show the filter input, a filterable prop wouldn't make much sense, so I chose showFilterInput

I see. I guess we could chalk this up to the different functionality then 👍

> I think the word "placeholder" is appropriate in your case. "My" props are "text"s as well. So the question is: label vs text (tbh I don't care) and prefix vs suffix. I def. prefer the term we choose to prefix the prop as that will group them in the prop types (and ideally in the component's argument list; that's not one of our conventions though).

In this case, I'd go with what the Select has done for simplicity. Unless you really hate it, but then we should also change it for the select.

> Two separate components is what I tried initially, it increased the complexity so I chose against it. This component does not have a single vs multi approach, instead the amount of selections can be limited. It currently only allows two values (1 and Infinity), but we can increase the range of allowed selections without introducing a breaking change, we'd just have to extend the UI. It also removed the need to handle two different prop types for selected. Bot the range instead of single/multi and the single prop type decrease complexity.

Makes sense, so yeah, then we can chalk this one up to a functionality difference as well 👍

Reply:
> In this case, I'd go with what the Select has done for simplicity. Unless you really hate it, but then we should also change it for the select.

I vote for going with the established way for now, and if it needs to be changed, raise an issue and propose how we do it across all components.

Reply:
I chose `selectedEmptyComponent` instead of `pickedEmptyComponent` as that matches with the `selected` prop